### ISSUE TITLE
Assertion type function not working as expected with everything but ES3 since >=4.9.4

### ISSUE THREAD

**Issue by carlo23666** [microsoft/TypeScript#52430]
# Bug Report

### 🔎 Search Terms

Assert, Assert Function, Assert Class Constructor

### 🕗 Version & Regression Information

`>=4.9.4`

~- This is a crash~
- This changed between versions 4.8.4 and >=4.9.4
- This is the behavior in every version I tried, and I reviewed the FAQ for entries about _________
- I was unable to test this on prior versions because it works as expected

### ⏯ Playground Link

https://www.typescriptlang.org/play?target=0&ts=5.0.0-dev.20230126#code/KYDwDg9gTgLgBDAnmYcCCAjAzjKBDAYxgBVlgAeYuAXjjwDtEA+GuACj210PnuAHd2AOhF4oAcywAuOowDaAXQCUNFsRUAyOADEArvSIBLCPQDcAKFCRYCMnADCJglGAxgpFJVYNmrNn0E2ESExSRkfRRVqNU0dfSMTCytoeCQUOA8KKlofFlpMHHwiTMoWAB8HJxc3EuImC0twFLhDejcoADNCVALuIjQsLGBYY3osOABvczgWsZgGAgoAUXBgInHQN3oAE3GS-QBregh+eiYmNmmZuiJdPAAbGUPj04AaK5mrNZhpOBWUdbvGZKcKDYY-G4wO73FrjACScwW7jI5H+3yw9XMAF9zI1rPACCYcHQwbAZL0ijABkMRkTWFMZq0cEiOLcHq84F91ioGddqrooPRIdDZsyDMAIB1Oat1hYZlj3liGvdXHAQKwAIwk2SIcx4UkwIRM+bitggDlLKBQaBKPUGo2I03muAAETwbiUQA

### 💻 Code

```ts
export type AbstractType<T = any> = (abstract new (...args: any[]) => T) & Function;
export type ConcreteType<T = any> = (new (...args: any[]) => T) & Function;
export type Type<T = any> = AbstractType<T> | ConcreteType<T>;

export interface AbstractAssertions {
  instance<Expects extends Type<unknown>>(
    actual: unknown,
    expects: Expects,
  ): asserts actual is InstanceType<Expects>;
}

export const assert: AbstractAssertions = {
  instance(actual, expects) {
    return actual instanceof expects;
  },
};

let x = 1 as any
assert.instance(x, Error) // Error here
assert.instance(x, Date) // Error here
```

### 🙁 Actual behavior

Show Type Error when not supposed to

### 🙂 Expected behavior

Assert type as expected

**Comment by MartinJohns** [comment]
Just pointing out that your `instance` implementation is not actually an assertion function (it throws), but a type guard (it returns a `boolean`).

**Comment by carlo23666** [comment]
> Just pointing out that your `instance` implementation is not actually an assertion function (it throws), but a type guard (it returns a `boolean`).

Yeah my bad, i forgot to delete some parts that might be unrelated to it haha

**Comment by RyanCavanaugh** [comment]
FYI @ahejlsberg - but I don't think this needs immediate fixing since the workaround is to remove an unneeded intersection

**Comment by ahejlsberg** [comment]
Turns out this is fixed by #52392. The issue is that we aren't properly propagating the `intersectionState` flags when checking signatures in type relations.

**Comment by ahejlsberg** [comment]
Ah, nevermind, it isn't fixed by #52392. However, it only reproduces when `--strictBindCallApply` is enabled.

**Comment by RyanCavanaugh** [comment]
This is fixed. With `--strictBindCallApply`, the original `assert.instance(x, Error)` and `assert.instance(x, Date)` calls report TS2345 in 5.4.0-dev.20231129 and compile in 5.4.0-dev.20231130.

The change is [#54753](https://github.com/microsoft/TypeScript/pull/54753), which also makes the original example compile in TypeScript 6.0.3 and the current native compiler.
